presto-main/src/main/java/io/prestosql/execution/buffer/SerializedPage.java

```diff
@@ -18,29 +18,35 @@
 
 import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
-import static io.prestosql.execution.buffer.PageCompression.COMPRESSED;
-import static io.prestosql.execution.buffer.PageCompression.UNCOMPRESSED;
+import static io.prestosql.execution.buffer.PageCodecMarker.COMPRESSED;
+import static io.prestosql.execution.buffer.PageCodecMarker.ENCRYPTED;
 import static java.util.Objects.requireNonNull;
 
 public class SerializedPage
 {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(SerializedPage.class).instanceSize();
-    private static final int PAGE_COMPRESSION_SIZE = ClassLayout.parseClass(PageCompression.class).instanceSize();
 
     private final Slice slice;
-    private final PageCompression compression;
     private final int positionCount;
     private final int uncompressedSizeInBytes;
+    private final byte pageCodecMarkers;
 
-    public SerializedPage(Slice slice, PageCompression compression, int positionCount, int uncompressedSizeInBytes)
+    public SerializedPage(Slice slice, byte pageCodecMarkers, int positionCount, int uncompressedSizeInBytes)
     {
         this.slice = requireNonNull(slice, "slice is null");
-        this.compression = requireNonNull(compression, "compression is null");
         this.positionCount = positionCount;
         checkArgument(uncompressedSizeInBytes >= 0, "uncompressedSizeInBytes is negative");
-        checkArgument(compression == UNCOMPRESSED || uncompressedSizeInBytes > slice.length(), "compressed size must be smaller than uncompressed size when compressed");
-        checkArgument(compression == COMPRESSED || uncompressedSizeInBytes == slice.length(), "uncompressed size must be equal to slice length when uncompressed");
         this.uncompressedSizeInBytes = uncompressedSizeInBytes;
+        this.pageCodecMarkers = pageCodecMarkers;
+        //  Encrypted pages may include arbitrary overhead from ciphers, sanity checks skipped
+        if (!ENCRYPTED.isSet(pageCodecMarkers)) {
+            if (COMPRESSED.isSet(pageCodecMarkers)) {
+                checkArgument(uncompressedSizeInBytes > slice.length(), "compressed size must be smaller than uncompressed size when compressed");
```

Comment:
Typically -- yes. In edge cases, the compressed data can be bigger than the original one though.

Reply:
The logic in PagesSerde will attempt compression but throw away the compressed result if ratio doesn't meet the minimum threshold. By the time the `SerializedPage` is created, the `uncompressedSize >= slice.length()` will always be true. The check was already present, but doesn't work with encrypted page since encryption adds an IV + padding to the input data (which could be compressed or not). While we could theoretically calculate the maximum possible overhead for AES and check against that, it seemed unnecessary to make `SerializedPage` aware of and I chose to say that encryption overhead is arbitrary and bail on the check instead.

Reply:
@pettyjamesm thanks, i missed the fact the check was already there